### PR TITLE
chore: analytics filters separation

### PR DESCRIPTION
**Pull request by sagarnaikjuspay**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
Currently Analytics amount value is invalid cause there's no currency seperation done so in this change will provide a seperate metrics for each currency. and also we decoupled the filters into saperate section. and graph UI enhancement
Screenshot 2024-06-25 at 5 32 39 PM
Screenshot 2024-06-25 at 5 32 47 PM

## Checklist

- [x] I ran `npm run re:build`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: src/components/DynamicChart.res [hunk 1]
```diff
@@ -737,7 +730,7 @@ let make = (
           <form onSubmit={handleSubmit}>
             <AddDataAttributes attributes=[("data-chart-segment", "Chart-1")]>
               <div
-                className="border rounded bg-white border-jp-gray-500 dark:border-jp-gray-960 dark:bg-jp-gray-950 dynamicChart pt-7">
+                className="border rounded-lg bg-white border-jp-gray-500 dark:border-jp-gray-960 dark:bg-jp-gray-950 dynamicChart pt-7">
                 {if chartLoading && shimmerType === Shimmer {
```
Comment (JeevaRamu0104) [inline src/components/DynamicChart.res:734]: do we need this shimmerType condition

File: src/components/HSwitchSingleStatTableWidget.res [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+  }
+}
+
+let percentFormat = value => {
+  `${Float.toFixedWithPrecision(value, ~digits=2)}%`
+}
+// if day > then only date else time
+let statValue = (val, statType) => {
+  open LogicUtils
+  if statType === "Amount" {
```
Comment (JeevaRamu0104) [inline src/components/HSwitchSingleStatTableWidget.res:55]: can we define enum for this value

**Review by JeevaRamu0104 — COMMENTED** [review]
image
please address these warnings from analytics module

File: src/screens/Analytics/AnalyticsNew.res [hunk 3]
```diff
@@ -0,0 +1,714 @@
...
+        })
+        dataDict->JSON.Encode.object
+      })
+      ->JSON.Encode.array
+      ->getTable
+      ->Array.map(Nullable.make)
+    }
+
+    open Promise
+    let getWeeklyData = async (data, cols) => {
```
Comment (JeevaRamu0104) [inline src/screens/Analytics/AnalyticsNew.res:269]: use async only when you want to use await

File: src/screens/Analytics/AnalyticsNew.res [hunk 4]
```diff
@@ -0,0 +1,714 @@
...
+          ~customFilter,
+          ~moduleName,
+          ~showDeltaMetrics=true,
+          (),
+        )
+
+        fetchDetails(tableEntity.uri, tableReqBody, Post, ())
+        ->thenResolve(json => {
+          switch weeklyTableMetricsCols {
+          | Some(cols) => getWeeklyData(json, cols)->ignore
```
Comment (JeevaRamu0104) [inline src/screens/Analytics/AnalyticsNew.res:327]: please recheck this function. try moving this out of useEffect and call as separate function.code is bit not readable

File: src/screens/Analytics/AnalyticsNew.res [hunk 5]
```diff
@@ -0,0 +1,714 @@
...
+            ->catch(_ => resolve())
+            ->ignore
+          | None => ()
+          }
+        } catch {
+        | _ => ()
+        }
+      }
+      None
+    }, (startTimeVal, endTimeVal, filterBody->JSON.Encode.object->JSON.stringify))
```
Comment (JeevaRamu0104) [inline src/screens/Analytics/AnalyticsNew.res:68]: api is getting called thrice if we apply some filters

https://github.com/juspay/hyperswitch-control-center/assets/120017870/1ebaab4a-4355-4822-87bd-f81c4b0e2982
try writing separate useEffect for filterbody

File: src/components/DynamicSingleStat.res [hunk 6]
```diff
@@ -282,9 +289,13 @@ let make = (
           source,
           prefix: ?urlConfig.prefix,
         }
-        let singleStatBodyMakerFn = urlConfig.singleStatBody->Option.getOr(singleStatBodyMake)
 
-        let singleStatBody = singleStatBodyMakerFn(singleStatBodyEntity)
+        let singleStatBodyMakerFn = switch formaPayload {
```
Comment (JeevaRamu0104) [inline src/components/DynamicSingleStat.res:293]: api is getting called twice if we apply some filters

https://github.com/juspay/hyperswitch-control-center/assets/120017870/1ebaab4a-4355-4822-87bd-f81c4b0e2982
image
image
Try moving this to separate function and call that function from useEffect

File: src/screens/Analytics/HomePageOverview/HomePageOverviewComponent.res [hunk 7]
```diff
@@ -103,11 +101,11 @@ module SystemMetricsInsights = {
     let defaultColumns: array<DynamicSingleStat.columns<systemMetricsSingleStateMetrics>> = [
       {
         sectionName: "",
-        columns: [Latency],
+        columns: [{colType: Latency}],
       },
     ]
 
-    let singleStatBodyMake = (singleStatBodyEntity: singleStatBodyEntity) => {
+    let _singleStatBodyMake = (singleStatBodyEntity: singleStatBodyEntity) => {
```
Comment (Riddhiagrawal001) [inline src/screens/Analytics/HomePageOverview/HomePageOverviewComponent.res:108]: If we are not using this func then can we remove it ?

File: src/screens/Analytics/PaymentsAnalytics/PaymentAnalytics.res [hunk 8]
```diff
@@ -44,6 +44,27 @@ let make = () => {
...
+    "avg_ticket_size",
+    "payment_processed_amount",
+  ]
+  let smartRetrieMetrics = ["retries_count"]
+
+  let formatMetrics = arrMetrics => {
+    arrMetrics->Array.map(metric => {
+      [("name", metric->JSON.Encode.string), ("desc", ""->JSON.Encode.string)]
+      ->Dict.fromArray
+      ->JSON.Encode.object
```
Comment (Riddhiagrawal001) [inline src/screens/Analytics/PaymentsAnalytics/PaymentAnalytics.res:64]: Please use `LogicUtils.getJsonFromArrayOfJson`

File: src/screens/Analytics/PaymentsAnalytics/PaymentAnalytics.res [hunk 9]
```diff
@@ -101,34 +119,125 @@ let make = () => {
...
+        ~groupByNames=["currency"]->Some,
+        ~customFilter=?singleStatBodyEntity.customFilter,
+        ~source=?singleStatBodyEntity.source,
+        ~granularity=singleStatBodyEntity.granularity,
+        ~prefix=singleStatBodyEntity.prefix,
+        (),
+      )->JSON.Encode.object,
+    ]
+    ->JSON.Encode.array
+    ->JSON.stringify
```
Comment (Riddhiagrawal001) [inline src/screens/Analytics/PaymentsAnalytics/PaymentAnalytics.res:143]: Please use LogicUtils.getJsonFromArrayOfJson

File: src/screens/Analytics/SystemMetrics/SystemMetricsAnalyticsUtils.res [hunk 10]
```diff
@@ -58,7 +56,14 @@ let defaultColumns: array<
...
     sectionName: "",
-    columns: [Latency, ApiCount],
+    columns: [
+      {
+        colType: Latency,
+      },
+      {
+        colType: ApiCount,
+      },
+    ],
```
Comment (Riddhiagrawal001) [inline src/screens/Analytics/SystemMetrics/SystemMetricsAnalyticsUtils.res:66]: Can we make a utility func which will generate this object based on the array passed ?

File: src/components/DynamicSingleStat.res (hunk 6 above)
Comment (sagarnaikjuspay) [reply src/components/DynamicSingleStat.res:293]: single state has a two api calls, one returns the single value and another returns a series value so when filters are changes two api calls happens, delta value is diff btw both the api calls

File: src/screens/Analytics/PaymentsAnalytics/PaymentAnalytics.res (hunk 9 above)
Comment (sagarnaikjuspay) [reply src/screens/Analytics/PaymentsAnalytics/PaymentAnalytics.res:143]: `LogicUtils.getJsonFromArrayOfJson` takes the format of `[(string,JSON)]` but here its just `[(JSON)]`

File: src/components/HSwitchSingleStatTableWidget.res [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+              newDesign=true
+            />
+          </div>
+          <div className="flex gap-1 flex-col w-full mt-1">
+            {if value->Array.length > 0 {
+              <>
+                {value
+                ->Array.filterWithIndex((_val, index) => index < 5)
+                ->Array.map(item => {
+                  <div className="flex justify-between w-full text-sm opacity-70">
```
Comment (JeevaRamu0104) [inline src/components/HSwitchSingleStatTableWidget.res:209]: add `key={index->Int.toString}`

File: src/screens/Analytics/AnalyticsNew.res [hunk 12]
```diff
@@ -0,0 +1,723 @@
...
+        activeTab->Option.getOr([])->Array.includes(val.key) ? Some(item) : None
+      })
+      ->Array.concat(allColumns)
+    }, [activeTabStr])
+
+    let transactionTableDefaultCols = React.useMemo2(() => {
+      Recoil.atom(`${moduleName}DefaultCols${activeTabStr}`, newDefaultCols)
+    }, (newDefaultCols, `${moduleName}DefaultCols${activeTabStr}`))
+
+    let timeRange =
```
Comment (JeevaRamu0104) [inline src/screens/Analytics/AnalyticsNew.res:377]: remove this default value

**Review by JeevaRamu0104 — APPROVED** [review]

**Pull request merged**